add design proposal template (#37)
cncf/project-template: DESIGN-PROPOSALS.md

## Risks and Mitigations

Describe the risks of this proposal and how they can be mitigated. This should be broadly scoped and describe how it will impact the larger ecosystem and potentially adopters of the project; such as if adopters need to immediately update, or support a new port or protocol. It should include drawbacks to the proposed solution. 

### Security Considerations

When attempting to identify security implications of the changes, consider the following questions:
* Does the change alter the permissions or access of users, services, components - this could be an improvement or downgrade or even just a different way of doing it?
* Does the change alter the flow of information, events, and logs stored, processed, or transmitted?
* Does the change increase the 'surface area' exposed - meaning, if an operator of the project or user were to go rogue or be uninformed in its operation, do they have more areas that could be manipulated unfavorably?

> * What existing security features, controls, or boundaries would be affected by this change? ```

Author: ```suggestion
* What existing security features, controls, or boundaries would be affected by this change?
```

# Design Proposal Template

**Authors**: alice alice@example.com, bob jones [bob@example.com](mailto:bob@example.com)

**Begin Design Discussion**: YYYY-MM-DD

> **(optional) Status: **implementable, accepted, deferred, rejected, withdrawn, or replaced

Reviewer: ```suggestion
**(optional) Status: **implementable, accepted, deferred, rejected, withdrawn, or replaced

**Checklist**:

- [ ] Prerequisite 1
- [ ] Prerequisite 2
- [ ] Component 1
- [ ] Component 2
- [ ] Docs
- [ ] Tests

```

We should have a checklist in here so that people can track how much of the proposal has been implemented.  It's part of the KEP template.

# Design Proposal Template

**Authors**: alice alice@example.com, bob jones bob@example.com

**Begin Design Discussion**: YYYY-MM-DD

**(optional) Status: **implementable, accepted, deferred, rejected, withdrawn, or replaced

**Checklist**:

> - [ ] Prerequisite 1
> - [ ] Prerequisite 2
> - [ ] Component 1
> - [ ] Component 2
> - [ ] Docs
> - [ ] Tests

Reviewer 2: ```suggestion
- [ ] Problem, need, impact, and solution definition
- [ ] User Story/Stories
- [ ] Proposal scoping (goals/non-goals)
- [ ] Detailed design, diagrams
- [ ] Impact, risk, and security considerations
- [ ] Observability, performance, and scaling considerations
- [ ] Availability, compatibility, and roll back considerations
- [ ] Implementation phases, planning timeline
- [ ] Test plan
- [ ] Accept/reject proposal
- [ ] Implementation
- [ ] Docs updates and release notes
- [ ] Release
```

I've suggested a more specific listing that aligns with the content of this doc. If this is more what you are looking for, i would recommend shifting the ordering of the doc to reflect the list here so authors can follow along (it'd be sweet if these were links to the sections of the doc, then it can be used as a Table of Contents )

Reviewer: Well, I was mostly trying to indicate that there should be some kind of checklist there, but what goes *in* the checklist is highly dependent on the project's own processes.  For example, KEPs have a very different-looking checklist: https://github.com/kubernetes/enhancements/issues/3668 .  I was trying NOT to have anything that suggested a specific process that probably wouldn't apply.

Author: I agree with Josh that this should be a checklist and not a table of contents.

Reviewer 2: Sounds good.  Providing instructions for this checklist would be helpful, particularly describing what it is intended to capture.

> - [ ] Tests

Reviewer 3: Should these "prerequisite" and "component" lines be something else?
I read @jberkus's comment to suggest that this checklist be all of the phases in the lifecycle of a design proposal, and they get checked off as the proposal moves through those phases. I assume that "prerequisite 1"... "Component 2" aren't phases in the lifecycle.

Reviewer 3: If the idea is for the authors to fill this checklist with pieces of the proposal and not lifecycle stages of a proposal, I think that we need to make that clear with a sentence or two above the checklist explaining how an author will use it.

Reviewer: It's supposed to be a generic "There should be a checklist here".  What goes in the checklist is highly dependent on how the individual project manages their PRs.

But regardless, I think we're getting to the point of "we can't move foward on this without also having documentation for the template user".

## Summary/Abstract

Provide a high-level summary of the proposed change. Keep it short.

> This design process is usually around adding a series of standards around recording change notes for the project. The intent is to make the process of building the release changelog faster and easier and track decisions for the future. It is usually a combination of a feature and effort-tracking document, a product requirements document, and a design document.

Reviewer 3: "recording change notes for the project" doesn't sound like what this is. I think that we need to clarify that this is for template for creating a design change or feature in a project and for tracking that as it gets implemented

## Background/Motivation

Provide the motivation for the feature and any context required to understand the motivation. The motivation should justify any potential impact and why it is important. Often these proposals start as an issue, forum post, email and it's helpful to link to it in this section to provide context and credit the right people for the idea.

It is vital for projects to be able to track the chain of custody for a proposed enhancement from conception through implementation which can sometimes be difficult to do in a single Github issue, especially when it is a larger design decision or cuts across multiple areas of the project.

The purpose of the design proposal processes is to reduce the amount of "siloed knowledge" in a community. By moving decisions from a smattering of mailing lists, video calls, slack messages, Github exchanges, and hallway conversations into a well tracked artifact, the process aims to enhance communication and discoverability.

## (Optional) User/User Story

> Detail the things that people will be able to do with the feature if it is implemented. Include as much detail as possible so that people can understand the "how" of the system. This can also be combined with the background/motivation above.

Reviewer 2: ```suggestion
Detail the things that people will be able to do with the feature if it is implemented. Include as much detail as possible so that people can understand the "how" of the system. This can also be combined with the prior sections.
```

Reviewer 3: I'd also add something about who the target users that will be using this feature are.

## Goals

> The desired goal and list goals that the design achieves. What is it trying to achieve? How will you know that it has succeeded?

Reviewer 2: ```suggestion
List the desired goal or goals that the design is intended to achieve. These goals can be leveraged to structure and scope the design and discussions and may be reused as the "definition of done" -  the criteria you use to know the implementation of the design has succeeded in accomplishing its goals.
```

## Non-Goals

Describe what is out of scope for the design proposal. Listing non-goals helps to focus discussion and make progress.

## Proposal

> This is where we get down to the specifics of what the proposal actually is. It should have enough detail that reviewers can understand exactly what you're proposing, but should not include things like API designs or implementation. Describe the desired outcome and how to measure success.

Reviewer 2: ```suggestion
This is where we get down to the specifics of what the proposal actually is. It should have enough detail that reviewers can understand exactly what you're proposing, but should not include things like API designs or implementation. This section should expand on the desired outcome and include details on how to measure success.
```

## Design Details

> This section should contain enough information that the specifics of the change are understandable. This may include API specs (though not always required) or even code snippets. If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them. This can also be combined with the proposal section above. It should also address how the solution is backward compatible and how to deal with these incompatibilities, possibly with defaulting or migrations.

Reviewer 2: ```suggestion
This section should contain enough information that the specifics of the change are understandable. This may include API specs (though not always required) or even code snippets. If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them. This can also be combined with the proposal section above. It should also address how the solution is backward compatible and how to deal with these incompatibilities, possibly with defaulting or migrations. It may be useful to refer back to the goals and non-goals to assist in articulating the "why" behind your approach.
```

Reviewer 3: I'd change "specifics" to something more, umm, specific. Like, should the section contain enough informaiton that the "proposed implementation" can be understood? Is this still for more depth to the product direction and not tech/implementation details?

## Impacts / Key Questions

List crucial impacts and key questions, some of which may still be open. They likely require discussion and are required to understand the trade-offs of the design. During the lifecycle of a design proposal, discussion on design aspects can be moved into this section. After reading through this section, it should be possible to understand any potentially negative or controversial impact of the design. It should also be possible to derive the key design questions: X vs Y.

This will also help people understand the caveats to the proposal, other important details that didn't come across above, and alternatives that could be considered. It can also be a good place to talk about core concepts and how they relate. It can be helpful to explicitly list the pros and cons of each decision

### Pros

### Cons

> 

Reviewer 2: ```suggestion
Cons are defined as the negative aspects or disadvantages of the design as described. This section has the potential to capture outstanding challenge areas or future improvements needed for the project and could be referenced in future PRs and issues. This is also a good place to check for any assumptions that have been made in the design.
```

## Risks and Mitigations

> Describe the risks of this proposal and how they can be mitigated. This should be broadly scoped and describe how it will impact the larger ecosystem. It should include drawbacks to the proposed solution and consider the security implications of the changes too.

Reviewer 2: ```suggestion
Describe the risks of this proposal and how they can be mitigated. This should be broadly scoped and describe how it will impact the larger ecosystem and potentially adopters of the project; such as if adopters need to immediately update, or support a new port or protocol. It should include drawbacks to the proposed solution and consider the security implications of the changes too. 

When attempting to identify security implications of the changes, consider the following questions:
* Does the change alter the permissions or access of users, services, components - this could be an improvement or downgrade or even just a different way of doing it?
* Does the change alter the flow of information, events, and logs stored, processed, or transmitted?
* Does the change increase the 'surface area' exposed - meaning, if an operator of the project or user were to go rogue or be uninformed in its operation, do they have more areas that could be manipulated unfavorably?
* What existing security features, controls, or boundaries would be affected by this change? ```

Reviewer 3: let's create a new section for security

This section can also be combined into the one above.

## (Optional) Future Milestones

List things that the design will enable but that are out of scope for now. This can help understand the greater impact of a proposal without requiring to extend the scope of a proposal unnecessarily.

## (Optional) Implementation Details

Some projects may desire to track the implementation details in the design proposal. Some sections may include:

### Testing Plan

An overview on the approaches used to test the implementation.

### Update/Rollback Compatibility

How the design impacts update compatibility and how users can test rollout and rollback.

### Scalability

Describe how the design scales, especially how changes API calls, resource usage, or impacts SLI/SLOs.

### Implementation Phases/History

> Detail is the design will get broken up into multiple phases and/or track the major milestones in the lifecycle of the design proposal.

Reviewer 2: ```suggestion
Describe the development and implementation phases planned to break up the work and/or record them here as they occur. Provide enough detail so readers may track the major milestones in the lifecycle of the design proposal and correlate them with issues, PRs, and releases occurring within the project.
```
